irohad/model/execution/impl/command_executor.cpp

```diff
@@ -309,62 +309,63 @@ namespace iroha {
+                                                WsvQuery &queries,
+                                                WsvCommand &commands) {
+      auto subtract_asset_quantity =
+          static_cast<const SubtractAssetQuantity &>(command);
 
-        auto asset = queries.getAsset(subtract_asset_quantity.asset_id);
-        if (not asset) {
-          log_->info("asset {} is absent", subtract_asset_quantity.asset_id);
-          return false;
-        }
-        auto precision = asset.value().precision;
+      auto asset = queries.getAsset(subtract_asset_quantity.asset_id);
+      if (not asset) {
+        log_->info("asset {} is absent", subtract_asset_quantity.asset_id);
+        return false;
+      }
+      auto precision = asset.value().precision;
 
-        if (subtract_asset_quantity.amount.getPrecision() != precision) {
-          log_->info("amount is wrongly formed");
-          return false;
-        }
-        auto account_asset = queries.getAccountAsset(
+      if (subtract_asset_quantity.amount.getPrecision() != precision) {
+        log_->info("amount is wrongly formed");
+        return false;
+      }
+      auto account_asset = queries.getAccountAsset(
           subtract_asset_quantity.account_id, subtract_asset_quantity.asset_id);
-        if (not account_asset.has_value()) {
-          log_->info("{} do not have {}",
-                     subtract_asset_quantity.account_id,
-                     subtract_asset_quantity.asset_id);
-          return false;
-        }
-        auto account_asset_value = account_asset.value();
+      if (not account_asset.has_value()) {
+        log_->info("{} do not have {}",
+                   subtract_asset_quantity.account_id,
+                   subtract_asset_quantity.asset_id);
+        return false;
+      }
+      auto account_asset_value = account_asset.value();
 
-        auto new_balance =
+      auto new_balance =
           account_asset_value.balance - subtract_asset_quantity.amount;
-        if (not new_balance.has_value()) {
-          return false;
-        }
-        account_asset->balance = new_balance.value();
-
-        // accountAsset.value().balance -= amount;
-        return commands.upsertAccountAsset(account_asset.value());
+      if (not new_balance.has_value()) {
+        return false;
       }
+      account_asset->balance = new_balance.value();
 
-      bool SubtractAssetQuantityExecutor::hasPermissions(const Command &command,
-                                                    WsvQuery &queries,
-                                                    const Account &creator) {
-        auto cmd_value = static_cast<const SubtractAssetQuantity &>(command);
-        return creator.account_id == cmd_value.account_id
-               and checkAccountRolePermission(
-          creator.account_id, queries, can_subtract_asset_qty);
-      }
+      // accountAsset.value().balance -= amount;
+      return commands.upsertAccountAsset(account_asset.value());
```

Comment:
Btw, didn't notice it at the first time. Afair optional has implicit cast to bool so it's replace this:
https://github.com/hyperledger/iroha/blob/a1cacb1ac9435fd2946423cc8d250ec0625281b5/irohad/model/execution/impl/command_executor.cpp#L343-L353
with this (though didn't test it):
```cpp
return new_account_asset_value->balance - subtract_asset_quantity.amount
    | [&account_asset](auto new_balance) {
  account_asset->balance = new_balance;
  return account_asset;
} | [](auto account_asset) {
  commands.upsertAccountAsset(account_asset);
}
```

But not really sure if it's better, decide on your own

Reply:
Although it looks nice, we still need to add logging for that. When valdiation fails the reason is not clear

Reply:
have we added it? @grimadas